src/mon/FSCommands.cc

```diff
@@ -1245,6 +1242,125 @@ class RenameFilesystemHandler : public FileSystemCommandHandler
   Paxos *m_paxos;
 };
 
+class SwapFilesystemHandler : public FileSystemCommandHandler
+{
+  public:
+  explicit SwapFilesystemHandler(Paxos *paxos)
+    : FileSystemCommandHandler("fs swap"), m_paxos(paxos)
+  {
+  }
+
+  bool batched_propose() override {
+    return true;
+  }
+
+  int handle(
+      Monitor *mon,
+      FSMap& fsmap,
+      MonOpRequestRef op,
+      const cmdmap_t& cmdmap,
+      std::ostream &ss) override
+  {
+    ceph_assert(m_paxos->is_plugged());
+
+    // Check for confirmation flag
+    bool confirmation_flag = false;
+    cmd_getval(cmdmap, "yes_i_really_mean_it", confirmation_flag);
+    if (!confirmation_flag) {
+      ss << "This is a potentially disruptive operation, client\'s cephx "
+	"credentials may need to be reauthorized to access the file systems "
+	"and its pools. Add --yes-i-really-mean-it if you are sure you wish "
+	"to continue.";
+      return -EPERM;
+    }
+
+    string fs1_name, fs2_name;
+    bool swap_flag;
+    cmd_getval(cmdmap, "fs1_name", fs1_name);
+    cmd_getval(cmdmap, "fs2_name", fs2_name);
+    cmd_getval(cmdmap, "swap_fscids", swap_flag);
+    auto fs1 = fsmap.get_filesystem(fs1_name);
+    auto fs2 = fsmap.get_filesystem(fs2_name);
```

Comment:
Thinking about this more: we need to rethink this API as it's not idempotent. It must be something like:

`ceph fs swap <fs1_name> <fs1_fscid> <fs2_name> <fs2_fscid> ...` 

Otherwise retrying the command due to loss of connectivity could reverse the desired outcome (i.e. restore the original names/ids).

Reply:
Or, we need to start maintaining state (somewhere) to ensure idempotent operation.

Reply:
I think what Patrick suggested in [first comment in this thread](https://github.com/ceph/ceph/pull/50212#discussion_r1172970988) is perfect. @batrick If you confirm, I'll add implement it.

Reply:
I think what Patrick suggested in [first comment in this thread](https://github.com/ceph/ceph/pull/50212#discussion_r1172970988) is perfect.

@batrick @vshankar If you confirm, I'll implement it.

Reply:
> Thinking about this more: we need to rethink this API as it's not idempotent. It must be something like:
> 
> `ceph fs swap <fs1_name> <fs1_fscid> <fs2_name> <fs2_fscid> ...`
> 
> Otherwise retrying the command due to loss of connectivity could reverse the desired outcome (i.e. restore the original names/ids).

This seems alright to me, however, the interface is not as nice as

> ceph fs swap fs1 fs2

:/

Reply:
How about:

> ceph fs swap fs@1 fs@2

?

Reply:
volume name cannot contain a `@`, so this is a tad bit better IMO.

Reply:
We could make `ceph.in` smarter by looking up the `fscid` for each fs and then passing the full swap command on. That's unprecedented I believe so I'm hesitant to advise it. This is such a special and rare command that I don't think it warrants a lot of effort to make it easy to use.

Reply:
Okay, cool.